Submit Homework 2 for Distributed Computing: MySQL cluster with GTID replication and data migration using Ansible and Docker

DmitryFedoroff opened the pull request:
- Replace single MySQL container with multi-node cluster including master and two slaves for data durability and scalability.
- Configure GTID-based replication with master.cnf, slave1.cnf, slave2.cnf and ensure synchronization via CHANGE REPLICATION SOURCE.
- Import legacy WordPress database data from old db_data volume into new cluster using temporary mysql_legacy container and docker exec pipeline.
- Ensure master and slaves are healthy with container info checks and validate data import before replication startup.
- Clean up legacy host_vars/local.yml and consolidate all configuration under group_vars/all.yml with replication credentials.
- Remove legacy templates wp-config.php.j2 and db-config.yml.j2 to rely on standard WordPress ENV configuration.
- Update Docker Compose services to include db-master, db-slave1, db-slave2 with fixed version tags, config mounts and healthchecks.
- Remove MySQL port binding to prevent external exposure and enforce internal-only communication via Docker network.
- Install Docker CE and Compose via apt using version-controlled strings for stable provisioning on remote Ubuntu host.
- Relocate nginx.conf to proper files/ path in docker_setup role and persist volumes across all major services.
- Reorganize site.yml and inventory to provision remote cloud instance using SSH with proper private key.
- Ensure WordPress starts only when database is healthy using healthcheck + depends_on with condition.
- Update visual documentation and replace outdated screenshots to reflect new architecture and configuration.

DmitryFedoroff committed:
build: Configure dependencies for container management

#comment Configure Ansible collection requirements for Docker-related modules ensuring community.docker availability

Affected: requirements.yml

DmitryFedoroff committed:
chore: Provide inventory and variable definitions

#comment Provide local connection settings, Python interpreter paths, and WordPress credentials for environment consistency

Affected: hosts.ini, local.yml, all.yml

DmitryFedoroff committed:
feat: Orchestrate site deployment

#comment Orchestrate Ansible roles for Docker, database, and WordPress to enable cohesive setup

Affected: site.yml

DmitryFedoroff committed:
feat: Establish Docker infrastructure

#comment Establish tasks for Docker Engine installation, Compose retrieval, and container orchestration with a dedicated network

Affected: main.yml, docker-compose.yml.j2

DmitryFedoroff committed:
feat: Configure database readiness

#comment Configure waiting mechanism for MySQL and specify credentials to ensure the service is ready before further steps

Affected: main.yml, db-config.yml.j2

DmitryFedoroff committed:
feat: Provision WordPress framework

#comment Provision tasks and templates for custom wp-config with essential database details and environment variables

Affected: main.yml, wp-config.php.j2

DmitryFedoroff committed:
feat: Configure Nginx routing

#comment Configure server block to route requests through PHP-FPM and serve the WordPress files from the correct document root

Affected: nginx.conf

DmitryFedoroff committed:
chore: Add screenshots folder for documentation

#comment Add graphical assets demonstrating deployment state including WordPress installation, admin console, public interface and running containers

Affected: screenshot_01.png, screenshot_02.png, screenshot_03.png, screenshot_04.png

DmitryFedoroff committed:
doc: Document deployment of containerized WordPress infrastructure

#comment Document project architecture, environment requirements, Ansible role structure, Docker service composition and startup procedure with embedded annotated screenshots for verification and reproducibility

Affected: README.md

DmitryFedoroff committed:
feat: Merge HW1 deployment pipeline for containerized WordPress infrastructure

This pull request finalizes Homework 1 for Distributed Computing by delivering a reproducible infrastructure-as-code solution for deploying WordPress using Ansible and Docker.

Highlights:
- Dockerized architecture for MySQL, WordPress (PHP-FPM), and Nginx.
- Ansible roles for Docker installation, service orchestration, and wp-config templating.
- Dynamic inventory and environment variable management.
- Custom Nginx configuration for PHP routing.
- Deployment validated via WordPress UI and CLI state (screenshots included).

DmitryFedoroff committed:
refactor: Update WordPress stack deployment for remote server compatibility and container stability

#comment Refactor WordPress stack deployment:

1. Remove legacy templates wp-config.php.j2 and db-config.yml.j2 to rely on built-in WordPress configuration via environment variables.
2. Configure Docker CE version 28.0.4 and docker-compose-plugin installation via apt with version validation, avoiding insecure curl-based downloads.
3. Relocate nginx.conf into role-specific files directory to follow Ansible structure conventions.
4. Eliminate redundant host_vars/local.yml and consolidate configuration under group_vars/all.yml.
5. Update MySQL, WordPress, and Nginx image tags to fixed stable versions for reproducibility and security.
6. Restrict MySQL service to internal Docker network by removing external port exposure.
7. Implement MySQL healthcheck and enforce startup dependency for WordPress container on database readiness.
8. Ensure persistent volumes are created for both database and application services.
9. Transition Ansible inventory to target remote Ubuntu host via SSH and update playbook to reflect new host group.
10. Maintain idempotency by avoiding shell or command tasks in favor of declarative Ansible modules.
11. Update deployment documentation and visual references via refreshed screenshots.

Affected: group_vars_all_yml, host_vars_local_yml, inventory_hosts_ini, requirements_yml, roles.db_setup.config_files_db_config_yml_j2, roles.db_setup.tasks_main_yml, roles.docker_setup.tasks_main_yml, roles.docker_setup.files_nginx_conf, roles.docker_setup.templates_docker_compose_yml_j2, roles.wordpress_setup.tasks_main_yml, roles.wordpress_setup.templates_wp_config_php_j2, screenshots_screenshot_01_png, screenshots_screenshot_02_png, screenshots_screenshot_03_png, screenshots_screenshot_04_png, site_yml

DmitryFedoroff committed:
refactor: Configure MySQL cluster with GTID replication and legacy data migration

#comment Configure multi-node MySQL cluster with one master and two replicas using GTID-based replication:

1. Replace single MySQL container with multi-node cluster including master and two slaves.
2. Create master.cnf, slave1.cnf and slave2.cnf with GTID, binlog, and replication directives.
3. Adjust docker-compose.yml to define db-master, db-slave1, db-slave2 with configuration volumes and healthchecks.
4. Integrate master/slave MySQL server configs into docker_setup role and copy them to remote host.
5. Detect legacy MySQL volume, run temporary container and dump all data into the new master.
6. Create replication user inside the master node using docker_container_exec.
7. Start replication on both slaves using CHANGE REPLICATION SOURCE TO with GTID-based auto-positioning.
8. Modify db_setup tasks to handle healthcheck, legacy container dump, and slave synchronization.
9. Refactor all.yml to include replication credentials and volumes for cluster migration.

Affected: group_vars/all.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2, docker_setup/files/master.cnf, docker_setup/files/slave1.cnf, docker_setup/files/slave2.cnf, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge pull request #2 from DmitryFedoroff/feat/hw2-distributed-computing-itmo

feat: Merge HW2 database clustering for MySQL GTID-based replication and legacy migration

DmitryFedoroff committed:
doc: Update README with GTID MySQL cluster architecture and migration flow

#comment Update documentation to reflect improvements in Homework 2 deployment pipeline:

1. Describe transition to distributed MySQL architecture with GTID replication (Master-Slave-Slave).
2. Document legacy volume detection and automated migration strategy via temporary container.
3. Outline Docker Compose network structure, healthchecks, and replication credentials setup.
4. Reflect infrastructure update to remote server with external Nginx port binding and container naming.
5. Clarify WordPress configuration without wp-config.php templating using environment variables.
6. Detail installation instructions including Python venv, Ansible Core, Galaxy collections, and SSH setup.
7. Include updated screenshots of admin UI, Docker containers, and WordPress landing page.
8. Present technical insights into failover, replication mechanics, and compose orchestration model.

Affected: README.md

DmitryFedoroff committed:
Merge pull request #3 from DmitryFedoroff/doc/update-readme-for-gtid-cluster-deployment

doc: Update README with GTID MySQL cluster architecture and migration flow

SemenMartynov commented:
Использование модулей `shell` и `command` нужно заменить на соответствующие модули, которые обеспечат декларативность.

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

SemenMartynov commented:
Так `docker_container_exec` это-ж тоже самое. Нужно использовать модули.

DmitryFedoroff commented:
> Так `docker_container_exec` это-ж тоже самое. Нужно использовать модули.

Я исходил из того, что применение `docker_container_exec` в сочетании с  `argv` и без `shell` это допустимый компромисс.

SemenMartynov commented:
Это перекладывание декларативного кода из одного модуля в другой...

SemenMartynov commented:
Я уточню свою мысль: бывают ситуации, когда и `shell` и `command` и `docker_container_exec` бывают оправданы.
Но для этого нужно обосновать такой выбор. Пока я вижу, что для многих задач, которые ты решаешь через shell, есть более подходящие модули.

DmitryFedoroff committed:
refactor: Transition to declarative MySQL cluster management with community.mysql

#comment Refactor infrastructure code to replace imperative docker_container_exec usage with safe and idempotent community.mysql modules for MySQL user creation, data import, and replication setup. Include dependency on community.mysql, enable PyMySQL, switch to declarative dump/import, and rework docker-compose for external network compatibility.

Affected: requirements.yml, db_setup/tasks/main.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2

DmitryFedoroff commented:
Удовлетворяют ли внесённые правки вашим требованиям? Могу ли я переходить к Заданию #3?

SemenMartynov commented:
Репликация не работает.

На мастере вижу таблицу `wordpress_db`, а на `mysql_slave1` её нет.

DmitryFedoroff committed:
refactor: Refine Docker and MySQL roles for idempotent secure replication

#comment Refactor installation and replication logic to increase idempotency, security and observability:

1. Collect package facts via apt and expose current docker‑ce and compose plugin versions for operators.
2. Verify version compliance against 5:28.0.4 and 2.20.0 thresholds and conditionally add Docker APT repo and GPG key.
3. Separate engine and plugin installation paths using docker_pkg_version and compose_plugin_min_version facts.
4. Skip redundant package upgrades when required versions already present to keep reruns fast and safe.
5. Recreate replication account with REQUIRE SSL and GTID‑safe grants to enforce encrypted channels.
6. Hard reset replica state on slaves and apply CHANGE PRIMARY with primary_ssl for auto‑positioned GTID replication.
7. Insert pause to allow catch‑up and assert IO and SQL thread health with explicit fail‑fast checks.
8. Validate wordpress_db existence on each replica and report Seconds_Behind_Source for troubleshooting.
9. Output informative debug messages to clarify play behaviour during dry‑runs and real executions.

Affected: docker_setup/tasks/main.yml, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge branch 'main' into feat/hw2-distributed-computing-itmo

SemenMartynov commented:
Уже можно смотреть?

DmitryFedoroff commented:
> Уже можно смотреть?

Да, исправления во 2-й лабораторной сделаны. В следующей 3-й лабораторной привожу весь проект к требуемой структуре

SemenMartynov reviewed (approved):
```
      - |
        DROP USER IF EXISTS '{{ replication_user }}'@'%';
      - |
        CREATE USER '{{ replication_user }}'@'%' IDENTIFIED BY '{{ replication_password }}';
```
Вместо этого, ты можешь сразу сделать "СОЗДАТЬ ЕСЛИ НЕ СУЩЕСТВУЕТ" и это будет более идемпотентный способ.

---

`/tmp/legacy_dump.sql`

Представь, что у тебя дамп содержит сотни гигабайт, или даже терабайты... В `tmp` места не хватит. Лучше путь для дампа задавать через переменную.

---

```
pause: seconds: 10
```

Это ненадёжный способ. А если 10 секунд не хватит? Лучше явно указать какого события ты ожидаешь, и проверять именно его.

---

Вообще, логика с репликацией не идемпотентна. Представь, что ты несколько раз дёргаешь один плейбук -- каждый раз кластер будет пересоздаваться, вместо того чтоб пропустить эти шаги.

---

Репликация работает.

```
mysql> SELECT * FROM test_table WHERE id = 1;
+----+---------------------+
| id | ts                  |
+----+---------------------+
|  1 | 2025-05-08 09:23:28 |
+----+---------------------+
1 row in set (0.00 sec)
```
